ruskeyverbs/views.py

```diff
@@ -26,19 +51,618 @@ def index(request):
 
 @login_required
 def VerbListPerUser(request):
-    # get list of all available verbs sorted by due date for given user
-    verbs_with_due_dates = Verb.objects.annotate(due=Min('example__performanceperexample__due_date', filter=Q(example__performanceperexample__user_id=request.user.pk))).order_by('due').exclude(due=None)
-    verbs_without_due_dates = Verb.objects.annotate(due=Min('example__performanceperexample__due_date', filter=Q(example__performanceperexample__user_id=request.user.pk))).exclude(~Q(due=None))
-    sorted_verb_list = list(chain(verbs_with_due_dates, verbs_without_due_dates))
-    # sorted_verb_list = sorted(Verb.objects.all(),
-    #                           key=lambda a: a.get_earliest_due_date(request.user))
-    current_user = request.user
+    """
+    list view for verbs available in the system. Sorts by due date
+    (if studied previously) then frequency ranking
+    """
+    # two query sets because otherwise unstudied verbs would display before
+    # studied ones
+    verbs_with_due_dates = Verb.objects.annotate(
+        due=Min(
+            'example__performanceperexample__due_date',
+            filter=Q(
+                example__performanceperexample__user_id=request.user.pk)
+            )
+        ).order_by('due').exclude(due=None)
+    verbs_without_due_dates = Verb.objects.annotate(
+        due=Min('example__performanceperexample__due_date',
+                filter=Q(
+                    example__performanceperexample__user_id=request.user.pk)
+                )
+        ).exclude(~Q(due=None))
+    sorted_verb_list = list(chain(verbs_with_due_dates,
+                                  verbs_without_due_dates))
     return render(
         request,
         'ruskeyverbs/list_view_per_user.html',
-        context={'verb_list': sorted_verb_list, 'current_user': current_user}
+        context={'verb_list': sorted_verb_list}
     )
 
 
-class VerbDetails(LoginRequiredMixin, generic.DetailView):
-    model = Verb
+@login_required
+def VerbDetails(request, pk):
+    """
+    detail view for verbs; page will display conjugation and example info/audio
+    for user
+    """
+    verb = Verb.objects.get(pk=pk)
+    example_pk = Example.objects.annotate(
+        due=Min('performanceperexample__due_date',
+                filter=Q(
+                    performanceperexample__user_id=request.user.pk)
+                )
+        ).filter(verb=verb).order_by('due')[0].pk
+    """
+    If the user has previously studied the example, send them right
+    into the quiz; otherwise, send them to the study page before the
+    quiz
+    """
+    try:
+        print(PerformancePerExample.objects.get(pk=example_pk,
+                                                user=request.user))
+        not_studied = False
+    except ObjectDoesNotExist:
+        not_studied = True
+
+    request.session['quiz_counter'] = 0
```

Comment:
I guess this is better than the user state table, but I still think these session values are unnecessary...

Reply:
I actually did manage to use the PerformancePerExample model for the functionality I was planning to put into user_state (i.e. determining which example should be studied next). Here, I am only using session variables to track a user's progress through a single quiz session, limit the length of that quiz session, and prevent the user from visiting quiz pages out of order. I was not planning to store the user's progress through a given quiz session because the quiz sessions are only 12 questions in length. Do you think there is a compelling reason to do so?

Reply:
Nope. That makes sense.